README.md

```diff
@@ -0,0 +1,133 @@
+
+Consult one of these implementations, or the Client docstring for example usage of the Client class.
+
+Aside from the actionlib SimpleActionServer, the Action Server holds an instance of the Task Manager.
+As the name suggests, this is the component that does the actual managing of the task (including its subtasks) for us.
+
+## Procedure
+
+Let's see what happens when a client sends a goal to the action server.
+Let's assume our client takes the high level natural language task *"Go to the kitchen, find a coke, and bring it to me."*
+With a good natural language parser, we can parse this to the following json object:
+```json
+{
+  "actions" :
+  [
+    {
+      "action": "navigate-to",
+      "object":
+      {
+        "type" : "room",
+        "name" : "kitchen"
+      }
+    },
+    {
+      "action": "find",
+      "object":
+      {
+        "type": "coke"
+      }
+    },
+    {
+      "action": "bring",
+      "object":
+      {
+        "type": "reference"
+      },
+      "target-location":
+      {
+        "type": "person",
+        "id": "operator"
+      }
+    }
+  ]
+}
+```
+This json object can be sent to the action server.
+
+### Configuration
+
+The server parses the object and passes the resulting Python dictionary (the `recipe`) to the task manager (`set_up_state_machine(recipe)`).
+The task manager then goes through the list of actions, instantiates actions and tries to chain their semantics.
+It does this by calling the `configure` method on every action.
+
+For example, the second action (*find a coke*) results in knowledge of a coke.
+We don't know anything about this object yet, but we expect that there will be a coke in the world model.
+This knowledge is part of the `ConfigurationResult` returned by `Find.configure(configuration_data)`.
+It is added as knowledge to the configuration data that is passed to the next action.
+The next action, (*bring it to me*), receives this knowledge, but does not necessarily need it.
+However, the `reference` (*it*) in the `bring` action configuration data lets the `bring` action know that there must be knowledge of an object that can be brought somewhere.
+Therefore, it will try to find this in the knowledge in its configuration data.
+In our example, this knowledge is available, so the `bring` action will grab the found coke and take it to the operator.
+When the required knowledge is not available, the action may return a ConfigurationResult specifying that information is missing.
+The server will notify the client of this result so that it can ask the user for more information.
+
+### Execution
+
+When all actions are successfully configured, the task manager is ready to start executing the actions.
+To do this, the action server calls `task_manager.execute_next_action()` while there are still remaining actions.
+It will return early if an action fails.
+In our example: if the `find` action fails (so no coke is found), the `bring` action will not be executed.
+
```

Comment:
My problem with the above two sections is that I have no goal, I don't know if I actually have to do or know all the above, or that it is just background information to understand where the Action I want to write fits in the picture.

In the latter case, it's a lot of detail information which at this point is unclear whether it is the Action Server doing this for me, or whether I have to do this myself in my action.

You may want to explain earlier that I am to write an action, possibly with a few lines of the kind of information that is needed for this, and the available resources for it. In that way, I can attach the details to the global picture, instead of having to remember it all without clear structure.

Reply:
Good idea. I'll swap the 'Implementing your own action' and 'Architecture' sections and say something about this to make it clearer.